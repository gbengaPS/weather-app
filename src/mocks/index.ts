export const citiesMock = {
  base: "stations",
  clouds: {all: 0},
  cod: 200,
  coord: {lon: 116.4, lat: 39.91},
  dt: 1602226045,
  id: 1816670,
  main: {temp: 23.52, feels_like: 22.22, temp_min: 23.33, temp_max: 23.89, pressure: 1014, humidity: 47},
  name: "Beijing",
  sys: {type: 3, id: 2000022, country: 'CN', sunrise: 1602195502, sunset: 1602236679},
  timezone: 28800,
  visibility: 10000,
  weather: [{id: 800, main: "Clear", description: "clear sky", icon: "01d"}],
  wind: {speed: 2.55, deg: 209},
};

export const initialStateMock =  {
  favorites: { 'Beijing': { data: citiesMock } },
  weatherInfo: [],
  notes: { 'Beijing': 'hello world' },
  isLoading: false
}