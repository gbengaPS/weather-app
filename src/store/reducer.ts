const reducer = (state: any, action: any) => {
  return {}
}

export default reducer;