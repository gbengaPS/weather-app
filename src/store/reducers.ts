import { REMOVE_WEATHER_INFO, ADD_WEATHER_INFO, ADD_FAVORITE, REMOVE_FAVORITE } from '../constants';
import { FavoritesMap, WeatherInfo } from '../types';
import { saveFavorites } from '../helpers';

export const weatherInfoReducer = (state: WeatherInfo[], action: any) => {
  switch(action.type) {
    case REMOVE_WEATHER_INFO:
      const filteredInfo = state.filter((info: WeatherInfo) => {
        return info.data.name !== action.cityName;
      });
      localStorage.setItem('largestCitiesInfo', JSON.stringify(filteredInfo))
      return filteredInfo;

    case ADD_WEATHER_INFO:
      return [ ...state, ...action.weatherInfo ];
    default:
      return state;
  }
}

export const favoritesReducer = (state: FavoritesMap, action: any) => {
  switch(action.type) {
    case ADD_FAVORITE:
      const newState = { ...state, ...action.favoriteItem };
      saveFavorites(newState)
      return newState;

    case REMOVE_FAVORITE:
      const stateCopy = {...state};
      delete stateCopy[action.cityName];
      saveFavorites(stateCopy);
      return stateCopy;
    default:
      return state
  }
}