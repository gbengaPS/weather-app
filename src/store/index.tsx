import React, { createContext, useReducer } from 'react';
import { FavoritesMap, WeatherInfo } from '../types';
import { weatherInfoReducer, favoritesReducer } from './reducers';

interface InitialStateType {
	favorites: FavoritesMap
	weatherInfo: WeatherInfo[]
}

const initialState  = {
	favorites: {},
	weatherInfo: []
}

const AppContext = createContext<{
    state: InitialStateType;
    dispatch: React.Dispatch<any>;
  }>({
    state: initialState,
    dispatch: () => null
  });

const mainReducer = (state: InitialStateType, action: any) => ({
  weatherInfo: weatherInfoReducer(state.weatherInfo, action),
  favorites: favoritesReducer(state.favorites, action)
});

const AppProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(mainReducer, initialState);

  return (
      <AppContext.Provider value={{state, dispatch}}>
      {children}
      </AppContext.Provider>
  )
}
  
  export { AppContext, AppProvider };