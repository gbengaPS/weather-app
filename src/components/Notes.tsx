import React, { useState, useContext, useEffect } from 'react';
import { AppContext } from '../store';
import { SAVE_NOTES, DELETE_NOTE } from '../constants';


const Notes = (props: {cityName: string}) => {
  const { state, dispatch } = useContext(AppContext);
  const { cityName } = props;
  const [notesEditable, setNotesEditable] = useState(false);
  const [notes, setNotes] = useState('');

  useEffect(() => {
    setNotes(state.notes[cityName] || '')
  }, [cityName, state.notes]);

  const handleSave = (event: any) => {
    event.preventDefault();
    setNotesEditable(false);
    dispatch({
      type: SAVE_NOTES,
      notes: { [cityName]: notes }
    });
  }
  
  const handleCommentChange = (event: any) => {
    setNotes(event.target.value)
  }

  const handleCancelClick = () => {
    setNotesEditable(false)
  }

  const handleDelete = () => {
    setNotesEditable(false);
    dispatch({
      type: DELETE_NOTE,
      cityName
    });
  }

  const cityNotes = state.notes[cityName];

  if (notesEditable) {
    return (
      <>
        <form>
          <textarea value={notes} onChange={handleCommentChange}></textarea>
        </form>
          <button className="notes-button" onClick={handleSave}>Save</button>
          <button className="notes-button" onClick={handleCancelClick}>Cancel</button>
          {
            cityNotes ? <button className="notes-button" onClick={handleDelete}>Delete</button> : null
          }
          
      </>
    )
  }

  return (
    <>
      { cityNotes ?  <p className="notes">{cityNotes}</p> : null }
      <button className="notes-button" onClick={() => setNotesEditable(true)} >
        { cityNotes ? 'Edit' : 'Add Notes' }
      </button>
    </>
  )
}

export default Notes;
