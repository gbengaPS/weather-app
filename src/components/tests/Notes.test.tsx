import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Notes from '../Notes';
import { AppContext } from '../../store';
import { SAVE_NOTES, DELETE_NOTE } from '../../constants';
import { initialStateMock as state } from '../../mocks';



describe('Notes', () => {

  const dispatch = jest.fn();

  it('should match snapshot', () => {
    const { container } = render(
      <AppContext.Provider value={{ dispatch, state }}>
        <Notes cityName="Beijing"/>
      </AppContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should show Add Notes button when notes are empty', () => {
    const { getByText } = render(
      <AppContext.Provider value={{ dispatch, state }}>
        <Notes cityName="Tokyo"/>
      </AppContext.Provider>
    );
    expect(getByText('Add Notes')).toBeInTheDocument();
  });

  it('should show textarea when Edit button is clicked', () => {
    const { getByText, getByTestId } = render(
      <AppContext.Provider value={{ dispatch, state }}>
        <Notes cityName="Beijing"/>
      </AppContext.Provider>
    );
    
    fireEvent.click(getByText('Edit'))
    
    expect(getByTestId('notes-input')).toBeInTheDocument();
    expect(getByText('Save')).toBeInTheDocument();
    expect(getByText('Cancel')).toBeInTheDocument();
  });

  it('should call dispatch when save button is clicked', () => {
    const { getByText, getByTestId } = render(
      <AppContext.Provider value={{ dispatch, state }}>
        <Notes cityName="Beijing"/>
      </AppContext.Provider>
    );
    
    fireEvent.click(getByText('Edit'));
    fireEvent.change(getByTestId('notes-input'), { target: { value: 'test' }})
    fireEvent.click(getByText('Save'));
    expect(dispatch).toHaveBeenCalledWith({
      type: SAVE_NOTES,
      notes: {'Beijing': 'test'}})
  });

  it('should call dispatch when delete button is clicked', () => {
    const { getByText } = render(
      <AppContext.Provider value={{ dispatch, state }}>
        <Notes cityName="Beijing"/>
      </AppContext.Provider>
    );
    
    fireEvent.click(getByText('Delete'));
    expect(dispatch).toHaveBeenCalledWith({
      type: DELETE_NOTE,
      cityName:'Beijing'
    });
  });

  it('should allow users cancel edits', () => {
    const { getByText, getByTestId } = render(
      <AppContext.Provider value={{ dispatch, state }}>
        <Notes cityName="Beijing"/>
      </AppContext.Provider>
    );

    expect(getByText('hello world')).toBeInTheDocument();

    fireEvent.click(getByText('Edit'));
    fireEvent.change(getByTestId('notes-input'), { target: { value: 'test' }})
    fireEvent.click(getByText('Cancel'));

    expect(getByText('hello world')).toBeInTheDocument();
  });
});
