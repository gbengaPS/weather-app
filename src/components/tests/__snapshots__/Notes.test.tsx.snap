// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes should match snapshot 1`] = `
<div>
  <div
    class="notes-wrapper"
  >
    <p
      class="notes"
    >
      hello world
    </p>
    <div
      class="action-buttons"
    >
      <button
        class="notes-button info"
      >
        Edit
      </button>
      <button
        class="notes-button danger"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
